core/alsp_src/library/rows_cols.pro

```diff
@@ -0,0 +1,567 @@
+	mk_rows_atoms(ListOfRows, NumCols, ListOfListsOfAtoms, 
+			ListOfItemsLens, ListOfMinColsWidthsLists),
+	final_cols_widths(ListOfMinColsWidthsLists, ColsWidths),
+	setup_rows_pads(ListOfItemsLens, ColsWidths, RowsPads),
+	!,
+	output_fixed_rows(ListOfListsOfAtoms, ColsWidths, RowsPads, OutputStream).
+
+/*!---------------------------------------------------------------------
+ |      columns/3
+ |      columns(ListOfRows, ColWidths, OutputStream)
+ |      columns(+, +, +) 
+ |
+ |      - Output ListOfRows as a columnized rectangular array of terms
+ |
+ |      If ListOfRows is a list of lists of terms (Rows), and
+ |	if all of the Rows are of the same length, and if ColWidths
+ |	is a list of positive integers of the same length as the 
+ | 	common length as the individual lists in ListOfRows, outputs a 
+ |      rectangular array of the terms appearing in the Rows to 
+ |	OutputStream. The array is split into columns of the size of
+ |	the corresponding integer of ColWidths, with one character-wide 
+ |	blank separators.  Terms in Rows which are of size > the 
+ |	corresponding integer of ColWidths are trucated.
+ |
+ | Examples:
+ |
+ |	?- listing(t2).
+ |	% user:t2/0
+ |	t2 :- silly_list(ListOfRows), columns(ListOfRows,[5,5,5,5]).
+ |	
+ |	?- t2.
+ |	abc   rjtir x     y     
+ |	yrtbv h8t   x     yacc  
+ |	34    f(4,5 some_ thru  
+ |	
+ |	?- listing(t2h).
+ |	% user:t2h/0
+ |	t2h :-  silly_list(ListOfRows),
+ |	        columns([u([apple,orange,banana,kiwi],0'=)|ListOfRows],[5,5,5,5]).
+ |	
+ |	?- t2h.
+ |	apple orang banan kiwi  
+ |	===== ===== ===== ===== 
+ |	abc   rjtir x     y     
+ |	yrtbv h8t   x     yacc  
+ |	34    f(4,5 some_ thru  
+ *!--------------------------------------------------------------------*/
+columns(ListOfRows, ColWidths, OutputStream)
+	:-
+	    % NumCols is a var, used to check all rows are the same length
+	fixed_mk_rows_atoms(ListOfRows, ColWidths, NumCols, RowsAtoms, RowsPads),
+	!,
+	output_fixed_rows(RowsAtoms, ColWidths, RowsPads, OutputStream).
+
+	/* ========= Variable column size support ========= */
+
+mk_rows_atoms([], NumCols, [], [], []).
+mk_rows_atoms([u(Row,Code) | RowsTail], NumCols, [RowAtoms, UAtoms | RowsAtomsTail], 
+	 	[RowItemsLen, RowItemsLen | RowsItemLensTail], 
+		[RowMinColWidths, RowItemsLen | RowsMinColsWidthsTail])
+	:-!,
+	mk_rows_atoms([Row | RowsTail], NumCols, [RowAtoms | RowsAtomsTail], 
+	 		[RowItemsLen | RowsItemLensTail], [RowMinColWidths | RowsMinColsWidthsTail]),
+	mk_u_atoms(RowItemsLen, Code, UAtoms).
+
+mk_rows_atoms([Row | RowsTail], NumCols, [RowAtoms | RowsAtomsTail], 
+	 [RowItemsLen | RowsItemLensTail], [RowMinColWidths | RowsMinColsWidthsTail])
+	:-
+	atoms_for_row(Row, 0, RowNumCols, RowAtoms, RowItemsLen, RowMinColWidths),
+	(RowNumCols = NumCols ->
+		mk_rows_atoms(RowsTail, NumCols, RowsAtomsTail, 
+				RowsItemLensTail, RowsMinColsWidthsTail)
+		;
+		throw(rows_len_error(Row,NumCols, RowNumCols))
```

Comment:
Yeah probably.  I was thinking about a last chance handler.
I won't disagree.  However, it's not clear to me what the ErrorTerm should be in:
        throw(error(ErrorTerm, ErrorInfo)).  
The wording of http://127.0.0.1:4000/docs/ref/catch.html (i.e., the list of ErrorTerms) suggests that user code can't throw errors??

Reply:
Uneven rows would be a `domain_error` (i.e. the domain of the function  `columns` only includes lists of eqi-length rows, just like `sqrt`'s domain is non-negative floats). You can use the `domain_error/3` helper function to throw the error.

I think it is ok for user code to throw errors, but in this case `columns` is a new builtin, so it should definitely throw appropriate errors like other builtins.

Reply:
I thought something like that at first too.  But the catch/3 page specifies:

domain_error(ValidDomain, Culprit) – The base type of an argument is correct, but the value is outside the domain for which the predicate is defined. The ISO Prolog Standard states that ValidDomain may be any one of the following atoms: character_code_list, character_list, close_option, flag_value, io_mode, not_less_than_zero, operator_priority, operator_specifier, prolog_flag, read_option, source_sink, stream_or_alias, stream_option, stream_position, write_option. Moreover, ALS Prolog allows ValidDomain to take on these additional values : depth_computation, line_length, positive_integer. Culprit is the argument which caused the error.

Unless I’m missing something, none of these domains would cover lists of eqi-length rows.   Where do we state that "sqrt's domain is non-negative floats” ?  I can’t find that, not in REf: is/2.   The domain_error extract above doesn’t list non-negative floats as an error_domain.

I’m not against adding more error domains.  In fact, non-negative floats for use wrt sqrt is probably important, and there are probably others.  And I think that a more “user-code-oriented” error domain ought to be thought about too.  It basically would be available for those situtations we haven’t been able to anticipate (e.g. eqi-length rows).

Also note the following non-error:

?- X is sqrt(-3.4).

X=0nan 

yes.

Maybe we should look into the current state of the ISO standard.  I know the committee still exists.

Reply:
Standard-wize, `ValidDomain` appears to be an open class -- ALS Prolog has already extended it, so the more the merrier.

Sorry about the confusion over `sqrt`. I was making a loose analogy (signaled by "just like"), but I should have said it in a purely mathematical sense:

> ... just like the mathematical function *square root*'s domain is the non-negative reals.
 
(of course, even here the analogy is loose because I don't specify that I'm talking about pre-Bombelli mathematics, which didn't have complex numbers).

The prolog `sqrt` function follows the IEEE floating point standard, which signals domain errors using NaNs.

Reply:
Resolved as: won't fix